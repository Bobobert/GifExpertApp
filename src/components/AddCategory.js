import React, { useState } from "react";
import PropTypes from "prop-types";

export const AddCategory = ( {setCategories} ) => {

    const [inputValue, setInputValue] = useState('');

    const handleInputChange = (e) => {
        setInputValue( e.target.value );
    }

    const handleSubmit = (e) => {
        e.preventDefault(); // avoids to refresh the page

        if (inputValue.trim().length > 1) {
            setCategories( categories => [inputValue, ...categories]);
            setInputValue('');
        }

    }

    return <form onSubmit={ handleSubmit }>
        <input 
            type="text"
            value={ inputValue }
            onChange={ handleInputChange }
        />
        </form>
}
//element from agroups every input element inside, so it works
// the same as a fragment in this snippet

AddCategory.propTypes = {
    setCategories : PropTypes.func.isRequired,
}