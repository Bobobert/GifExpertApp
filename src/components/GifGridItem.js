import React from 'react';

export const GifGridItem = ({id, title, url}) => {

    const handleCopy = () => {
        navigator.clipboard.writeText(`https://giphy.com/embed/${id}`).then(
        console.log(`Copied to clipboard ${title} !`));
    }

    return (
        <div className="card animate__animated animate__fadeInDown animate__fast" onClick={handleCopy}>
            <img src={ url } alt={ title }/>
            <div className='overlay'>
                <div className='text'>
                    This GIF!<br/>{title}
                </div>
            </div>

        </div>
    )
}
// extracted more info and style from this https://www.w3schools.com/css/css3_images.asp
