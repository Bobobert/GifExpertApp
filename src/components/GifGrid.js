import React from 'react'
import { GifGridItem } from './GifGridItem';
import { useFetchGifs } from '../hooks/useFetchGifs';
// import { getGifs } from '../helpers/getGifs';

export const GifGrid = ({ category }) => {

    //const [images, setImages] = useState([]);
    const { data:images, loading } = useFetchGifs( category );

    // useEffect(() => {
    //     getGifs();
    // }, []); // este snip sirve para bloquear la ejecucion de funciones
    // cuando son ejecutadas/ renderiadas por primera vez, al menos.
    // El problema estÃ¡ en que todo  este componente se ejecuta cuando se realiza
    // un cambio.

    // useEffect( () => {
    //     getGifs( category ).then( setImages );
    // }, [ category ]); // run just once for []
    // // or when category changes [ category ]
    
    return <>
            <h3> { category } </h3>
            { loading && <p className='animate__animated animate__shakeY'>Cargando . . .</p> }
            <div className='card-grid'>
                {images.map( img => 
                    <GifGridItem key={img.id}{...img} />
                    )}
            </div>
        </>
}
