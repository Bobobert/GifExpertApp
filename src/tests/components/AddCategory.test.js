import { render, screen } from "@testing-library/react";
import "@testing-library/jest-dom";
import React from "react";

import { AddCategory } from "../../components/AddCategory";
import userEvent from "@testing-library/user-event";

describe('Pruebas en <AddCategory />', () => {

    const setCategoriesTest = jest.fn(); 
    // este es como una simple funcion vacia () => {} pero mejor
    let container = null;

    beforeEach(() => {
        const {container: temp} = render(<AddCategory setCategories={setCategoriesTest}/>);
        container = temp;
    })

    afterEach(() => {
        jest.clearAllMocks();
    })

    test('debe mostrarse correctamente', () => {

        expect(container).toMatchSnapshot();

    })

    test('debe capturar correctamente el input', () => {

        const inputTest = 'Hallo world'
        const input = screen.getByRole('textbox');
        //userEvent.type(input, inputTest); // direct way to do it
        userEvent.keyboard(inputTest); 
        // textbox not selected? input not recorded
        // added autoFocus to the input!

        expect( input ).toHaveValue( inputTest );
    })

    test('no debe de postear la forma', () => {

        userEvent.keyboard('{enter}'); // it works! :o
        expect( setCategoriesTest ).not.toHaveBeenCalled();

    })

    test('debe llamar el setCategories y limpiar la caja de texto', () => {
        
        userEvent.keyboard('Super test{enter}');
        // expect( setCategoriesTest ).toHaveBeenCalled();
        expect( setCategoriesTest ).toHaveBeenCalledTimes( 1 );
        expect( screen.getByRole('textbox') ).toHaveValue('');

    })
})