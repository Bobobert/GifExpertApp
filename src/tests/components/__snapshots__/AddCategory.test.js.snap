// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en <AddCategory /> debe mostrarse correctamente 1`] = `
<div>
  <form>
    <input
      type="text"
      value=""
    />
  </form>
</div>
`;
